src/Conversion/ONNXToKrnl/Math/Gemm.cpp

```diff
@@ -47,13 +47,12 @@ struct ONNXGemmOpLowering : public ConversionPattern {
         [&](KrnlBuilder &createKrnl, ValueRange outerIndices) {
           // Create temp and set to zero, single scalar, no need for default
           // alignment.
-          MemRefBuilder createMemRef(createKrnl);
-          MathBuilder createMath(createKrnl);
-          Value red = createMemRef.alloca(MemRefType::get({}, elementType));
+          MultiDialectBuilder<MemRefBuilder, MathBuilder> create(createKrnl);
+          Value red = create.mem.alloca(MemRefType::get({}, elementType));
```

Comment:
In this case, I am thinking it looks natural to support `create.krnl` because `create` was created from `createKrnl`. It's helpful so we support it.

Reply:
That is exactly what I wanted to do, and I think I know how to do it :-)

Reply:
@tungld After further reflection, I don't think its possible. The entire scheme works on template specialization, meaning that between the
```
template<Ts...>
struct MultiDialectBuilder<MathBuilder, Ts>
```

and 
```
template<Ts...>
struct MultiDialectBuilder <MemRefBuilder, Ts>
```

I can generate a different specialization of `MultiDialectBuilder`, one that includes respectively the `math` object member and the other one the `men` object member.

However, the type used in the constructor cannot be used for specialization, its essentially a type of one of the function. This is not part of the signature type of the overall class, and thus I cannot specialize the type for it. 

Another difficulty is that all subclasses, such as `ONNXBuilder` and `KrnlBuilder` are not defined in the MLIRDialectBuilder; specialization can be added though, and it works, but I cannot "add" all of the variants upfront in the MultDialectBuilder.

So, we I believe we  must add each of the desired builder explicitly. 

I listed below a small program that I used to test the scheme. Feel free to experiment
```C++
#include <stdio.h>

class Top {};

class A: public Top {
public:
  A(int i) : val(i) {}
  A(A &aa) : val(aa.val) {}

  int getVal() { return val; }

private:
  int val;
};

class B: public Top {
public:
  B(int i) : val(10 + i)  {}
  B(B &bb) : val(bb.val) {}
  int getVal() { return val; }

private:
  int val;
};

class C: public Top {
public:
  C(int i) : val(100 + i) {}
  C(C &cc) : val(cc.val) {}
  int getVal() { return val; }

private:
  int val;
};

template <class... Ts>
class Builder {
public:
  Builder(int i) {}
};

template <class... Ts>
class Builder<A, Ts...> : public Builder<Ts...> {
public:
  Builder(int i) : Builder<Ts...>(i), a(i) {}
  A a;
};

template <class... Ts>
class Builder<B, Ts...> : public Builder<Ts...> {
public:
  Builder(int i) : Builder<Ts...>(i), b(i) {}
  B b;
};

template <class... Ts>
class Builder<C, Ts...> : public Builder<Ts...> {
public:
  Builder(int i) : Builder<Ts...>(i), c(i) {}
  C c;
};

int main() {
#if 1
  Builder<B, A, C> builder(1);
#else
  Builder<A, B, C> builder(1);
#endif

  int aa = builder.a.getVal();
  printf("aa is %d\n", aa);
  int bb = builder.b.getVal();
  printf("bb is %d\n", bb);
  int cc = builder.c.getVal();
  printf("cc is %d\n", cc);
  return 1;
}
```

Reply:
@etiotto maybe you have an idea if @tungld idea is feasible. Happy to hear ideas